## Problem 9: AI-Assisted Programming
Use ChatGPT or any other large language model (LLM) to generate a function called multivariate normal density(x, mu, Sigma) which returns the density of a D-dimensional vector x given a D-dimensional mean (location) vector
mu and a D×D-dimensional covariance matrix Cov. Compare the outputs of your
function with those obtained using SciPy’s scipy.stats.multivariate_normal
for a few parameterizations including a spherical Gaussian (zero covariance, shared variance across dimensions), a diagonal Gaussian (zero covariance, different variance for each dimension), and a full-covariance Gaussian (non-zero covariance, different variance for each dimension). Describe briefly how the LLM performed.


In [2]:
# Important imports
import numpy as np
import scipy.stats as stats

### AI-Generated Function:

In [21]:
def multivariate_normal_density(x, mu, Sigma):
    """
    Calculate the density of a D-dimensional vector x given a D-dimensional mean (location) vector mu
    and a covariance matrix Sigma.

    Parameters:
    - x: D-dimensional numpy array representing the vector whose density needs to be calculated.
    - mu: D-dimensional numpy array representing the mean vector.
    - Sigma: DxD numpy array representing the covariance matrix.

    Returns:
    - density: Density of the vector x.
    """
    D = len(mu)
    det = np.linalg.det(Sigma)
    inv_Sigma = np.linalg.inv(Sigma)
    norm_const = 1.0 / np.sqrt((2 * np.pi) ** D * det)
    x_mu = x - mu
    exponent = -0.5 * np.dot(np.dot(x_mu, inv_Sigma), x_mu.T)
    density = norm_const * np.exp(exponent)
    return density


### Spherical Gaussian
For the spherical Gaussian, we use a covariance matrix with zero covariance and the same variance for each dimension.

In [37]:
# Parameters
D = 3  # Number of dimensions
variance = 2.0
mu = np.random.rand(D)
Sigma = np.eye(D) * variance  # Spherical covariance matrix
x = np.random.rand(D)

# Calculate densities
ai_density = multivariate_normal_density(x, mu, Sigma)
scipy_density = stats.multivariate_normal(mean=mu, cov=Sigma).pdf(x)

density_difference = np.abs(ai_density - scipy_density)

# Compare
print(f"Spherical Gaussian:\nAI-Generated function density: {ai_density}\nSciPy density: {scipy_density}\nDifference: {density_difference}")

Spherical Gaussian:
AI-Generated function density: 0.020130695075512038
SciPy density: 0.020130695075512045
Difference: 6.938893903907228e-18


### Diagonal Gaussian
For the diagonal Gaussian, we use a covariance matrix with zero covariance and different variances for each dimension.

In [28]:
# Parameters
variances = np.random.rand(D) * 3  # Different variances
Sigma = np.diag(variances)  # Diagonal covariance matrix
x = np.random.rand(D)

# Calculate densities
ai_density = multivariate_normal_density(x, mu, Sigma)
scipy_density = stats.multivariate_normal(mean=mu, cov=Sigma).pdf(x)

density_difference = np.abs(ai_density - scipy_density)

# Compare
print(f"Diagonal Gaussian:\nAI-Generated function density: {ai_density}\nSciPy density: {scipy_density}\nDifference: {density_difference}")

Diagonal Gaussian:
AI-Generated function density: 0.025491653953254187
SciPy density: 0.025491653953254197
Difference: 1.0408340855860843e-17


### Full-Covariance Gaussian
For the full-covariance Gaussian, we use a covariance matrix with non-zero covariance and different variances for each dimension.

In [38]:
# Parameters
Sigma = np.random.rand(D, D)
Sigma = np.matmul(Sigma, Sigma.T) # Ensure the covariance matrix is positive semi-definite
x = np.random.rand(D)

# Calculate densities
ai_density = multivariate_normal_density(x, mu, Sigma)
scipy_density = stats.multivariate_normal(mean=mu, cov=Sigma).pdf(x)

density_difference = np.abs(ai_density - scipy_density)

# Compare
print(f"Full-Covariance Gaussian:\nAI-Generated function density: {ai_density}\nSciPy density: {scipy_density}\nDifference: {density_difference}")


Full-Covariance Gaussian:
AI-Generated function density: 2.153809648055261
SciPy density: 2.1538096480551
Difference: 1.6076029396572267e-13


### LLM Performance Discussion
#### Describe briefly how the LLM performed.
The AI-generated function's performance in calculating the density of a multivariate normal distribution is notably precise, closely mirroring the results from the well-established SciPy library, although not exactly matching them. The exceedingly small differences in the output, such as `6.94 x 10^-18` for the Spherical Gaussian and slightly larger but still minimal discrepancies for the Diagonal and Full-Covariance Gaussians, highlight a level of accuracy that surpasses typical expectations for AI-generated algorithms. These minor variances, while potentially attributable to the nuances of floating-point computations, importantly display that the AI's output is not an exact replica of SciPy's function.